pkg/scheduler/objects/application.go

```diff
@@ -93,6 +97,7 @@ type Application struct {
 	execTimeout          time.Duration               // execTimeout for the application run
 	placeholderTimer     *time.Timer                 // placeholder replace timer
 	gangSchedulingStyle  string                      // gang scheduling style can be hard (after timeout we fail the application), or soft (after timeeout we schedule it as a normal application)
+	startTime            time.Time                   // the time that the application starts running. Default is zero.
```

Comment:
One question, when we are starting state the app for gang scheduler enabled, it means the gang scheduler placeholder allocations are all allocated, if these placeholder allocations need to calculate the resource usage, because before the startTime here, the gang scheduler allocations are all bound in k8s shim side.
And if the gang scheduler timeout, if the placeholder allocations need to calculate even the app failed, because some placeholder allocations have been occupied resources for some time, and the default timeout is 15 mins, it's not a short time for resources occupied.

cc @wilfred-s @craigcondit

Reply:
Hi Qi, thanks for the thoughts. My understanding is that placeholder allocations as treated as being used, see 
    func (sa *Application) removeAllocationInternal(
only when the placeholder allocations are replaced with "real" allocations, they will be considered used.
I may have misunderstanding here, further comments are welcome.
Thanks,

Reply:
Thanks @craigcondit 

I have some questions about the startTime.

Currently I implemented the startTime in the app state machine as: 

```
fmt.Sprintf("enter_%s", Starting.String()): func(_ context.Context, event *fsm.Event) {
				app := event.Args[0].(*Application) //nolint:errcheck
				app.startTime = time.Now()
```
				
I guess the Start state in this diagram (https://yunikorn.apache.org/docs/next/design/scheduler_object_states/) means the scheduler started working on allocation for the app. Wonder if it makes sense to have a state in the diagram to represent the first pod binding? If there is such a state, we can set the startTime in the state machine when entering this state.

Otherwise, I studied the code and saw that we can possibly have a field bindingTime in the Allocation structure to represent the binding time, and set it in the following method:

`func (sa *Application) tryNodes(ask *AllocationAsk, iterator NodeIterator) *Allocation {`

would you please confirm?

Thanks.

Reply:
If you're going to account for usage accurately, you need to track each task from start to finish, not at the app level. If you have 1000 tasks, but can only run 10 at a time, you're going to have a stair-step chart as the first batch of 10 run and complete, then the next 10, etc. If each task takes 1 minute to complete, an accurate assessment of total time should be 1000 minutes. But tracking each task from the start of the application, you will account for 1 minute for each of the first 10 pods, 2 minutes for each of the next 10, etc. The final batch of 10 (100 batches total) will be tracked as 100 minutes each. On average, 50 minutes will be accounted to each task, so you will over-report actual usage by a factor of 50.

Reply:
The right way to do this would be to track the timestamp per allocation when the allocation is associated with a node, and when an allocation terminates, report that usage as ((allocationEndTime - allocationStartTime) * resources.

Reply:
Thanks @craigcondit , that boils down to the last question in my previous comment:

Otherwise, I studied the code and saw that we can possibly have a field bindingTime in the Allocation structure to represent the binding time, and set it in the following method:

`func (sa *Application) tryNodes(ask *AllocationAsk, iterator NodeIterator) *Allocation {`

I think the allocationEndTime of your term is the time when the allocation is released, which I currently use and the pod finish time.

Would you please advice?

Thanks.

Reply:
Yes, tracking the time (per allocation) between allocated and released will work. The application doesn't really impact this at all.

Reply:
A future enhancement could be to track total accumulated resources per unit time in the application for released allocations, and then provide an API to also get "in-progress" allocations. This could be used to get better granularity than task termination for long-running applications.

Reply:
Thanks, agreed.

Reply:
Hi Craig,

I found that we can not recover the bindTime for recovered allocation because si.Allocation come from the shim, unless we change the protocol of for this structure:

// Create a new Allocation from a node recovered allocation.
// Also creates an AllocationAsk to maintain backward compatible behaviour
// This returns a nil Allocation on nil input or errors
func NewAllocationFromSI(alloc *si.Allocation, instType string) *Allocation {

Or we simply set the current time to bind time for this kind of allocation? It seems that if the recovered allocation is because of the yunikorn restart, then we do need to let the shim to pass this info.

Thanks.

Reply:
Hi @craigcondit Another question, I saw that all callers of the following method pass "Reserved" as the value for the first parameter, wonder if it's intended to pass a different value at all, given that it's meant to be a reserved allocation.

```
func newReservedAllocation(result AllocationResult, nodeID string, instType string, ask *AllocationAsk) *Allocation {
	alloc := NewAllocation("", nodeID, instType, ask)
```
Thanks.